BPF based NodePort

borkmann opened the pull request:
Rebased all of #8256 into fresh PR. Needs #8382 merged first, and dropped off from here. Also depends on #8314 to avoid NAT cleanup panic. /cc @brb

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8384)
<!-- Reviewable:end -->

brb committed:
bpf: Extend ct_state to include node_port flag

The node_port flag is going to indicate that a flow is to an
endpoint via (k8s) NodePort service.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
bpf: Add support for local NodePort

This commit introduces a BPF support for (k8s) NodePort for cases
when an service endpoint is running on the same host which a client
sent a request to. Currently, IPv4-only.

This requires the following:

1. For the bpf_netdev ingress program the LB lookup is added.
2. CT entries are created in a global CT map for NodePort traffic.
   CT_EGRESS entry is marked with "node_port=1" so that bpf_lxc can
   distinguish such traffic later on.
3. bpf_lxc creates CT_INGRESS entry with "node_port=1" (an extra lookup
   is required).
4. Instead of doing reverse DNAT translation, bpf_lxc redirects
   a reply packet to bpf_netdev.
5. bpf_netdev on its ingress path does the reverse DNAT translation,
   changes src and dst mac addrs after consulting with the FIB, and
   finally sends the packet back.

In addition, we restructured the bpf_netdev.c code in a way that
the MASQ path is moved from its section to the common one and
is enabled based on whether ENABLE_MASQUERADE is defined.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Add --enable-node-port flag

The flag enables NodePort in BPF.

Currently, we do not populate NodePort services based on k8s
events.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
bpf: only bother with actual nodeport range

Don't attempt to do the lookup if the range is not for node port
services since we *only* expose node port for the phys netdev, but
not regular services.

Also exclude RELATED since not needed here.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
daemon: implicitly enable host services when node port is enabled

Given we share the same services table, we can reuse all of the host
services load balancing facility to get connectivity working from a
Cilium host:

  # ./daemon/cilium-agent --auto-direct-node-routes=true --ipv4-range 10.11.0.0/16 --kvstore-opt consul.address=192.168.1.125:8500 --kvstore consul --tunnel=disabled --device=eno1 --enable-ipv6=false --disable-envoy-version-check=true --masquerade=false --enable-node-port=true

  # cilium endpoint list
  ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                                              IPv6   IPv4            STATUS
             ENFORCEMENT        ENFORCEMENT
  868        Disabled           Disabled          17521      container:maintainer=NGINX Docker Maintainers <docker-maint@nginx.com>          10.11.9.146     ready
  3486       Disabled           Disabled          4          reserved:health                                                                 10.11.134.107   ready

  # cilium service list
  ID     Frontend              Backend
  3000   192.168.1.125:31000   1 => 10.12.111.220:80   <-- node port service on remote node
  3001   192.168.1.125:31001   1 => 10.11.9.146:80     <-- node port service on local node

  # curl 192.168.1.125:31000
  <!DOCTYPE html>
  <html>
  <head>
  <title>Welcome to nginx!</title>
  [...]

  # curl 192.168.1.125:31001
  <!DOCTYPE html>
  <html>
  <head>
  <title>Welcome to nginx!</title>
  [...]

tcpdump for the remote node:

  # tcpdump -i any port 80 or port 31000 -n
  tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
  listening on any, link-type LINUX_SLL (Linux cooked), capture size 262144 bytes
  14:45:08.951028 IP 192.168.1.125.45096 > 10.12.111.220.80: Flags [S], seq 427343062, win 64240, options [mss 1460,sackOK,TS val 788775526 ecr 0,nop,wscale 7], length 0
  14:45:08.951284 IP 10.12.111.220.80 > 192.168.1.125.45096: Flags [S.], seq 3138813364, ack 427343063, win 65160, options [mss 1460,sackOK,TS val 884710867 ecr 788775526,nop,wscale 7], length 0
  14:45:08.951327 IP 192.168.1.125.45096 > 10.12.111.220.80: Flags [.], ack 1, win 502, options [nop,nop,TS val 788775526 ecr 884710867], length 0
  14:45:08.951388 IP 192.168.1.125.45096 > 10.12.111.220.80: Flags [P.], seq 1:84, ack 1, win 502, options [nop,nop,TS val 788775526 ecr 884710867], length 83: HTTP: GET / HTTP/1.1
  14:45:08.951636 IP 10.12.111.220.80 > 192.168.1.125.45096: Flags [.], ack 84, win 509, options [nop,nop,TS val 884710867 ecr 788775526], length 0
  14:45:08.951659 IP 10.12.111.220.80 > 192.168.1.125.45096: Flags [P.], seq 1:239, ack 84, win 509, options [nop,nop,TS val 884710868 ecr 788775526], length 238: HTTP: HTTP/1.1 200 OK
  14:45:08.951673 IP 192.168.1.125.45096 > 10.12.111.220.80: Flags [.], ack 239, win 501, options [nop,nop,TS val 788775527 ecr 884710868], length 0
  14:45:08.951678 IP 10.12.111.220.80 > 192.168.1.125.45096: Flags [P.], seq 239:851, ack 84, win 509, options [nop,nop,TS val 884710868 ecr 788775526], length 612: HTTP
  14:45:08.951688 IP 192.168.1.125.45096 > 10.12.111.220.80: Flags [.], ack 851, win 501, options [nop,nop,TS val 788775527 ecr 884710868], length 0
  14:45:08.951863 IP 192.168.1.125.45096 > 10.12.111.220.80: Flags [F.], seq 84, ack 851, win 501, options [nop,nop,TS val 788775527 ecr 884710868], length 0
  14:45:08.952061 IP 10.12.111.220.80 > 192.168.1.125.45096: Flags [F.], seq 851, ack 85, win 509, options [nop,nop,TS val 884710868 ecr 788775527], length 0
  14:45:08.952083 IP 192.168.1.125.45096 > 10.12.111.220.80: Flags [.], ack 852, win 501, options [nop,nop,TS val 788775527 ecr 884710868], length 0

One thing that still needs to be resolved for this case is that the node
port must be published for all local addresses, e.g. also 127.0.0.1.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: wild card lookup for node port services from host

Now, instead of populating concrete IPs in our node port services
range, we can use wild-cards instead that the Cilium agent must
propagate to the services map.

Thus prior example ...

  # cilium service list
  ID     Frontend              Backend
  3000   192.168.1.125:31000   1 => 10.12.111.220:80   <-- node port service on remote node
  3001   192.168.1.125:31001   1 => 10.11.9.146:80     <-- node port service on local node

... now becomes:

  # cilium service list
  ID     Frontend              Backend
  3000   0.0.0.0:31000         1 => 10.12.111.220:80
  3001   0.0.0.0:31001         1 => 10.11.9.146:80

This is useful because it allows to enable node port for a Cilium
host for *all* local addresses:

  # curl 127.0.0.1:31000
  <!DOCTYPE html>
  <html>
  <head>
  <title>Welcome to nginx!</title>
  [...]

  # curl 192.168.1.125:31000
  <!DOCTYPE html>
  <html>
  <head>
  <title>Welcome to nginx!</title>
  [...]

  # curl 127.0.0.1:31001
  <!DOCTYPE html>
  <html>
  <head>
  <title>Welcome to nginx!</title>
  [...]

  # curl 192.168.1.125:31001
  <!DOCTYPE html>
  <html>
  <head>
  <title>Welcome to nginx!</title>
  [...]

Also, this prevents from conflicting IPs that would use the same node-port
port number which should not be possible.

For bpf_netdev which is attached to the phys device, using a wild card lookup
would be desirable, but not worth the complexity right now. Reason is that
doing the revnat would always result in a 0 source address, so this would
need some additional mapping elsewhere. Therefore we do it only for the host
where this doesn't matter.

Hence service table with mixed local + public exposure looks as follows:

  # cilium service list
  ID     Frontend              Backend
  3001   0.0.0.0:31001         1 => 10.11.9.146:80     <-- host local only
  3002   192.168.1.125:31001   1 => 10.11.9.146:80     <-- eno1 device

This actually has the advantage that for public exposure we can have more
fine-grained control.

The limitation with this apprach is however, that for connections to addresses
not managed by Cilium at all the node-port dst port range always tries to
select a backend for the host / pods.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf, nat: parameterize nat target range for reuse

We want to use it for normal masquerade but also for upcoming node to
node node-port communication, therefore we must pass target addr and
port range around. No functional change. Also remove most of the
ENABLE_MASQUERADE since we need these functions under node-port as
well, let the compiler evict.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add support for node to node node-port

This implements node-port support for the case where the endpoint is non-local.
This means traffic arriving on the node needs to hair pin the requests back out
to the backend on the final target node and performing BPF-based SNAT with its
own IP address. The reply path needs to perform the reverse NAT, redirecting the
reply back to the client.

This completes initial data path node-port support now covering all cases:

  - external traffic reaches host, EP is on local node
  - external traffic reaches host, EP is on remote node
  - internal, local traffic on host, EP is on local node
  - internal, local traffic on host, EP is on remote node

  # ./daemon/cilium-agent --auto-direct-node-routes=true --ipv4-range 10.11.0.0/16 --kvstore-opt consul.address=192.168.1.125:8500 --kvstore consul --tunnel=disabled --device=eno1 --enable-ipv6=false --disable-envoy-version-check=true --masquerade=false --enable-node-port=true

tcpdump example of an external client reaching the node-port service where the
backend is on the remote node (178.199.11.62 is client, 10.12.111.220 is remote
backend):

On 192.168.1.125 node:

  [...]
  IP 178.199.11.62.39528 > 192.168.1.125.31000: Flags [S], seq 934536923, win 64240, options [mss 1460,sackOK,TS val 1588108864 ecr 0,nop,wscale 7], length 0
  IP 192.168.1.125.33021 > 10.12.111.220.80: Flags [S], seq 934536923, win 64240, options [mss 1460,sackOK,TS val 1588108864 ecr 0,nop,wscale 7], length 0
  IP 10.12.111.220.80 > 192.168.1.125.33021: Flags [S.], seq 4149115260, ack 934536924, win 65160, options [mss 1460,sackOK,TS val 1236247213 ecr 1588108864,nop,wscale 7], length 0
  IP 192.168.1.125.31000 > 178.199.11.62.39528: Flags [S.], seq 4149115260, ack 934536924, win 65160, options [mss 1460,sackOK,TS val 1236247213 ecr 1588108864,nop,wscale 7], length 0
  [...]

  # cilium service list
   ID     Frontend              Backend
   3002   192.168.1.125:31000   1 => 10.12.111.220:80    <-- node port service on remote node

On 192.168.1.120 node containing the 10.12.111.220 backend:

  [...]
  IP 192.168.1.125.32899 > 10.12.111.220.80: Flags [S], seq 939783878, win 64240, options [mss 1460,sackOK,TS val 1658693390 ecr 0,nop,wscale 7], length 0
  IP 10.12.111.220.80 > 192.168.1.125.32899: Flags [S.], seq 3490274372, ack 939783879, win 65160, options [mss 1460,sackOK,TS val 1306831740 ecr 1658693390,nop,wscale 7], length 0
  [...]

On 178.199.11.62 client:

  # curl 192.168.1.125:31000
  <!DOCTYPE html>
  <html>
  <head>
  <title>Welcome to nginx!</title>
  [...]

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: full data path ipv6 support for node-port

Add IPv6 support for all connectivity cases for node-port in direct routing:

  # ./daemon/cilium-agent --auto-direct-node-routes=true --ipv4-range 10.11.0.0/16 --ipv6-range 2001:db8:3::/96 --kvstore-opt consul.address=192.168.1.125:8500 --kvstore consul --tunnel=disabled --device=eno1 --enable-ipv4=true --enable-ipv6=true --disable-envoy-version-check=true --masquerade=false --enable-node-port=true

Service table on node 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925:

  # cilium service list
  ID     Frontend                                         Backend
  3005   [2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925]:31000   1 => [2001:db8:3::346a]:80   <-- local
  3006   [2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925]:31001   1 => [2001:db8:4::f422]:80   <-- remote
  3007   [::]:31000                                       1 => [2001:db8:4::f422]:80   <-- wildcard

Curling from local node:

  # curl [::1]:31000
  <html><body><h1>It works!</h1></body></html>

Curling from remote node:

  # curl [2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925]:31000
  <html><body><h1>It works!</h1></body></html>

  # curl [2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925]:31001
  <html><body><h1>It works!</h1></body></html>

Node-port service when backend is on same node:

  [...]
  IP6 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f90f.50124 > 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925.31000: Flags [S], seq 2140201688, win 64800, options [mss 1440,sackOK,TS val 811210248 ecr 0,nop,wscale 7], length 0
  IP6 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f90f.50124 > 2001:db8:3::346a.80: Flags [S], seq 2140201688, win 64800, options [mss 1440,sackOK,TS val 811210248 ecr 0,nop,wscale 7], length 0
  IP6 2001:db8:3::346a.80 > 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f90f.50124: Flags [S.], seq 3156863179, ack 2140201689, win 64260, options [mss 1440,sackOK,TS val 790228078 ecr 811210248,nop,wscale 7], length 0
  IP6 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925.31000 > 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f90f.50124: Flags [S.], seq 3156863179, ack 2140201689, win 64260, options [mss 1440,sackOK,TS val 790228078 ecr 811210248,nop,wscale 7], length 0
  [...]

Node-port service when backend is on different node (via BPF SNAT):

  [...]
  IP6 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f90f.37604 > 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925.31001: Flags [S], seq 2251406673, win 64800, options [mss 1440,sackOK,TS val 812563108 ecr 0,nop,wscale 7], length 0
  IP6 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925.32899 > 2001:db8:4::f422.80: Flags [S], seq 2251406673, win 64800, options [mss 1440,sackOK,TS val 812563108 ecr 0,nop,wscale 7], length 0
  IP6 2001:db8:4::f422.80 > 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925.32899: Flags [S.], seq 1891482566, ack 2251406674, win 64260, options [mss 1440,sackOK,TS val 3278817564 ecr 812563108,nop,wscale 7], length 0
  IP6 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f925.31001 > 2a02:120b:c3fc:feb0:ec4:7aff:fe03:f90f.37604: Flags [S.], seq 1891482566, ack 2251406674, win 64260, options [mss 1440,sackOK,TS val 3278817564 ecr 812563108,nop,wscale 7], length 0
  [...]

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: proper error handling for drop notifications

Make errors visible for cilium monitor and properly handle tc action
as dropping.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb committed:
bpf: Enable NAT with ENABLE_{MASQUERADE,NODEPORT} conditions

Otherwise, bpf_netdev.c and bpf_alignchecker.c won't compile
when masq or nodeport is disabled.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
bpf: work around verifier issue in __ct_update_timeout

Program load fails with LLVM 7.y/8.y's code generation in the CT's
__ct_update_timeout():

  [...]
  level=warning msg="334: (b7) r1 = 48" subsys=daemon
  level=warning msg="335: (7b) *(u64 *)(r10 -280) = r1" subsys=daemon
  level=warning msg="336: (85) call bpf_ktime_get_ns#5" subsys=daemon
  level=warning msg="337: (37) r0 /= 1000000000" subsys=daemon
  level=warning msg="338: (0f) r8 += r0" subsys=daemon
  level=warning msg="339: (63) *(u32 *)(r9 +32) = r8" subsys=daemon
  level=warning msg=" R0_w=inv(id=0) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8_w=inv(id=0) R9=map_value(id=0,off=0,ks=14,vs=56,imm=0) R10=fp0,call_-1 fp-8=??mmmmmm fp-16=mmmmmmmm fp-24=m??????? fp-64=????mmmm fp-72=mmmmmmmm fp-80=mmmmmmmm fp-120=????0000 fp-128=??mmmmmm fp-136=mmmmmmmm fp-144=????mmmm fp-152=00000000 fp-160=00000000 fp-168=00000000 fp-176=00000000 fp-184=00000000 fp-192=00000000 fp-200=00000000 fp-208=00000000 fp-216=mmmmmmmm fp-224=map_value fp-232=mmmmmmmm fp-240=map_value fp-248=00000000 fp-256=mmmmmmmm fp-264=mmmmmmmm fp-272=mmmmmmmm fp-280=mmmmmmmm" subsys=daemon
  level=warning msg="340: (bf) r1 = r9" subsys=daemon
  level=warning msg="341: (79) r2 = *(u64 *)(r10 -280)" subsys=daemon
  level=warning msg="342: (0f) r1 += r2" subsys=daemon
  level=warning msg="math between map_value pointer and register with unbounded min value is not allowed" subsys=daemon
  [...]

LLVM 6.y seems fine and no error occurs. The offset seems to be stored
on stack by LLVM which is causing the issue. Work around it via READ_ONCE/
WRITE_ONCE helper. Also the indirection via pointer is removed.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb committed:
iptables: Disable MASQ for NodePort if BPF NodePort enabled

This commit disables the MASQUERADE rule for NodePort/HostPort if
NodePort support is enabled in BPF datapath. We can disable it, as
the case covered by the rule is handled by the transparent
host-lb which does not require SNATing.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
loadbalancer: Add L3n4AddrID.Equals() method

The method is going to be used for checking NodePort frontend
equality in the upcoming commits.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
k8s: Add NodePorts field to Service struct

The field is going to store NodePort frontends for a given service
of the NodePort type.

loadbalancer.L3n4AddrID{} was chosen to store a NodePort frontend
because the frontend will get a service ID allocated later by
k8s_watcher.go which will use the ID for reverse NAT xlation and
for deleting the related service.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Handle NodePort services

This commit extends the service management functionality by adding
a support for NodePort services.

Each NodePort service entry is stored in the same BPF map as ClusterIP
service. The main difference between those two is that for NodePort
service we create two frontends - one with a zero IP addr, another with
a public interface (`--device=<..>`) IP addr.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
bpf: skip pinning calls/policy tail call map

Otherwise ownership will potentially have a program type mismatch.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -203,6 +203,10 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 	case CT_REPLY:
 		policy_mark_skip(skb);
 
+#ifdef ENABLE_NODEPORT
```
It would be good to add this to the `bpf/node_config.h` so that we always compile this code in CI (and run the verifier tests against this code).

joestringer replied:
Note: On tip of this branch I hit this with it always enabled, not sure why exactly:

```
$ make -C bpf
  CC    bpf/bpf_lb.ll  [-DSKIP_DEBUG]
  CC    bpf/bpf_lxc.ll  [-DSKIP_DEBUG]
  CC    bpf/bpf_sock.ll  [-DSKIP_DEBUG]
  CC    bpf/bpf_netdev.ll
bpf_netdev.c:94:31: error: unused function 'tc_index_skip_nodeport' [-Werror,-Wunused-function]
static inline bool __inline__ tc_index_skip_nodeport(struct __sk_buff *skb)
                              ^
1 error generated.
make: *** [Makefile.bpf:23: bpf_netdev.ll] Error 1
make: *** Waiting for unfinished jobs....

```

borkmann replied:
Fixed, will push out in a bit.

joestringer commented on bpf/bpf_netdev.c:
```diff
@@ -243,6 +504,222 @@ static inline __u32 derive_ipv4_sec_ctx(struct __sk_buff *skb, struct iphdr *ip4
+
+/* Main node-port entry point for host-external ingressing node-port traffic
+ * which handles the case of: i) backend is local EP, ii) backend is remote EP,
+ * iii) reply from remote backend EP.
```
:+1: Super helpful comment.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -511,6 +515,12 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 
+#ifdef ENABLE_NODEPORT
+		if (ct_state.node_port) {
+			return redirect(NATIVE_DEV_IFINDEX, 0);
```
nit: The commit message in step (4) doesn't describe which direction this goes, but this would be going to the TC egress BPF program of the native device if I'm not mistaken (since it's `0`, not `BPF_F_INGRESS`)?

Step (5) in the commit message then describes how on `bpf_netdev` ingress path it does reverse DNAT, but it seems like the `rev_nodeport_lb4()` call is actually happening in the `to_netdev` (egress) program.

borkmann replied:
Yes, tc egress, this is only for the case where the EP is local to the node, and hit in reply path. I'll add a comment.

joestringer commented on bpf/bpf_netdev.c:
```diff
@@ -97,6 +116,224 @@ static inline __u32 derive_sec_ctx(struct __sk_buff *skb, const union v6addr *no
+	case CT_NEW:
+		ct_state_new.src_sec_id = SECLABEL;
+		ct_state_new.node_port = 1;
+		ret = ct_create6(get_ct_map6(&tuple), &tuple, skb, CT_EGRESS,
```
Should we consider splitting nodeport CT into its own map, to at least have the failure domain separate from cluster-internal connectivity? The type of thing I'm concerned about is the ability for someone to spam the nodeport and disrupt pod-to-pod connectivity by bumping CT entries. Arguably you can put mitigations in place for this using things like the XDP prefilter, but even so I think this is worth considering. Would there be drawbacks to such an approach? I guess that it may complicate the bpf_lxc egress path (Would need to either have an additional nodeport map lookup in `bpf_lxc` egress or otherwise inform the local CT map for this entry that it's nodeport, eg by marking a bit on the packet from `bpf_netdev` ingress prog and then storing it with the local CT map entry when sending the traffic into the endpoint)

borkmann replied:
Still brainstorming on this one. On the one hand it makes sense to me to have these separate, at least for node-external traffic. Internal nodeport traffic is handled via cgroups hooks, so it looks like regular connections to the data path, but it's also not too relevant in this context given other EP on the same node could spam the CT as well. Problem with external nodeport traffic where backend EP is on the same node is that the two CT map lookups will cause verifier complexity explosion in bpf_lxc and also sit in critical fast-path. Somehow it would be nice to have the same single LRU map and have different eviction policies, e.g. all internal traffic would be in one zone and all external traffic would be marked with a different one, so updates with a zone flag only ever evict entries in their own zone, problem somewhat is that this sizing needs to be declared upfront. Perhaps we should also make the initial backend selection random for a TCP connection for the case of nodeport traffic under the assumption that backends reside on several nodes anyway so the load gets distributed.

joestringer reviewed (commented):
I barely scratched the first couple of commits, but already had a few discussion items so figured I'll send out the review. I do intend to come back to this again tomorrow.

borkmann committed:
bpf: enable nodeport for compilation tests

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer commented on bpf/lib/conntrack.h:
```diff
@@ -309,9 +316,10 @@ static inline int __inline__ ct_lookup6(void *map, struct ipv6_ct_tuple *tuple,
+	int ret = CT_NEW, ret2 = CT_NEW, action = ACTION_UNSPEC;
 	bool is_tcp = tuple->nexthdr == IPPROTO_TCP;
 	union tcp_flags tcp_flags = { .value = 0 };
+	struct ct_state ct_state_tmp = {};
```
nit: These new fields can have a smaller scope, they're only used in the new CT lookup case below. No need to zero-initialize a `ct_state` struct if we don't hit that case.

joestringer commented on bpf/lib/conntrack.h:
```diff
@@ -423,6 +431,26 @@ static inline int __inline__ ct_lookup6(void *map, struct ipv6_ct_tuple *tuple,
+		 * flag value. The flag is set in the endpoint -> client
+		 * TUPLE_F_OUT entry which is created by bpf_netdev on its
+		 * ingress hook. */
+		if (ret == CT_NEW && dir == CT_INGRESS) {
```
Do we need this extra lookup in all cases for new connections, for instance `lb4_local()` case, or the new ones from `bpf_netdev`?

If I follow correctly, the time we need to fetch this bit is when we received inbound nodeport traffic (for a local endpoint) from outside the node for the first time. Just wondering if we could persist a bit across the stack forwarding somehow to propagate this info and avoid the extra lookup.

Or alternatively, skip this extra lookup from codepaths that don't need this information.

aanm commented on daemon/daemon_main.go:
```diff
@@ -615,6 +615,9 @@ func init() {
 	flags.String(option.LB, "", "Enables load balancer mode where load balancer bpf program is attached to the given interface")
 	option.BindEnv(option.LB)
 
+	flags.Bool(option.EnableNodePort, false, "Enable NodePort type services by Cilium")
```
1) Why aren't we enable it by default?
2) If we can't enable it by default because it's beta maybe we should add it in the option description `"Enable NodePort type services by Cilium (beta)"`

joestringer replied:
AFAICT We can't support BPF-based nodeport on all kernel versions that Cilium supports, so that may play into the "should we set it by default" question.

aanm replied:
We already enable / disable certain features based on the kernel versions.

borkmann replied:
Yes, we can flip the switch once we have everything integrated, and when we detect (probes, on TODO) that kernel fully supports it.

borkmann replied:
Unless we cannot get this from k8s, I'll add the range parameter, makes sense.

joestringer commented on pkg/datapath/linux/config.go:
```diff
@@ -169,27 +182,41 @@ func (l *linuxDatapath) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeConf
 	}
 
+	if option.Config.EnableNodePort {
+		fmt.Fprintf(fw, "#define ENABLE_NODEPORT 1\n")
```
This at minimum requires `HAVE_FIB_LOOKUP` (which we've already got a probe for), possibly other helpers too; How are we signalling to users when their kernel is too old for BPF nodeport?

borkmann replied:
Probes are still on todo list also for host based LB, so they will be combined together. We'll test for recvmsg hook as an indicator once it hits stable. Added to tracker.

aanm commented on pkg/k8s/service.go:
```diff
@@ -91,12 +93,44 @@ func ParseService(svc *types.Service) (ServiceID, *Service) {
+				proto := loadbalancer.L4Type(port.Protocol)
+				port := uint16(port.NodePort)
+				id := loadbalancer.ID(0) // will be allocated by k8s_watcher
+
```
to avoid calculating the String() 4 times you can create a variable with the string already here `nodePortFEStr := nodePortFE.String()`

borkmann replied:
As discussed offline, string is each time different (v4/v6 and wild-card/non-wild-card), so okay as is.

joestringer commented on bpf/bpf_netdev.c:
```diff
@@ -97,6 +116,224 @@ static inline __u32 derive_sec_ctx(struct __sk_buff *skb, const union v6addr *no
+	ipv6_addr_copy((union v6addr *) &fib_params.ipv6_src, (union v6addr *) &ip6->saddr);
+	ipv6_addr_copy((union v6addr *) &fib_params.ipv6_dst, (union v6addr *) &ip6->daddr);
+
+	ret = fib_lookup(skb, &fib_params, sizeof(fib_params),
```
nit: Might be tidier to declare our own `fib_lookup4(skb, iphdr,  ifindex)` / `fib_lookup6(skb, iphdr, ifindex)` functions that do the `revalidate_data()`, `fib_params` init, helper call and mac rewrite, since there's a few places this kind of code is used these days. Could be a follow-up.

aanm reviewed (changes requested):
@borkmann it seems that you have defined consts for min and max values of nodeports. K8s allows the node-ports to be set as an option so it would be nice to have that option as well in `cilium-agent`
`--service-node-port-range`

joestringer commented on bpf/bpf_netdev.c:
```diff
@@ -112,6 +349,17 @@ static inline int handle_ipv6(struct __sk_buff *skb, __u32 src_identity)
+#ifdef ENABLE_NODEPORT
+	if (!tc_index_skip_nodeport(skb) &&
+	    nodeport_lb6(skb, src_identity) < 0) {
+		return DROP_INVALID;
```
`nodeport_lb6()` returns a real error code, this is hiding it.

joestringer commented on bpf/bpf_netdev.c:
```diff
@@ -256,6 +733,17 @@ static inline int handle_ipv4(struct __sk_buff *skb, __u32 src_identity)
 
+#ifdef ENABLE_NODEPORT
+	if (!tc_index_skip_nodeport(skb) &&
+	    nodeport_lb4(skb, src_identity) < 0) {
```
`nodeport_lb4()` returns a real error code, this is hiding it.

joestringer commented on bpf/lib/nat.h:
```diff
@@ -104,7 +104,14 @@ struct ipv4_nat_entry {
+	const __be32 addr;
+	const __u16 min_port; /* host endianess */
+	const __u16 max_port; /* host endianess */
+	const bool force_range;
```
It'd be helpful to add a comment here explaining briefly what `force_range` is intended to do. (Ensure that the nat port is selected within the specified range on the specified address?)

joestringer reviewed (changes requested):
I had a few other pieces of feedback below.

Also, the final commit message states that the nodeport service entry is stored with two frontends - one with zero, and one with the IP of the native device. However I recall earlier commits stating that we would support nodeport on any local IP address. Is this just because only the native device ingress program will run the nodeport code (ie, support applying nodeport to traffic coming in)? Or will it also be attached to the `cilium_host` device, in which case do we also need a frontend for the IP configured there? I guess in most cases for local traffic, the bpf_sock program handles supporting nodeport for all local traffic. But what if one node directs traffic towards another node's internal IP (eg on `cilium_host`) with the nodeport? Then it would tunnel to that destination, route towards `cilium_host` and ...?

A few more general things:
* Is L7 entirely orthogonal to this logic? For instance if someone runs a nodeport in front of a local service that has some HTTP policy. I guess that in this case, the nodeport translation occurs in the native device, then the packet is redirected and arrives at either `cilium_host` or the lxc device, at which point the regular Cilium path takes over (and directs the traffic to the proxy). But the response to the initial connection from the proxy would go.... directly back out the native device, not through `cilium_host`? I don't see any reason why this should break, just wanted to check if this case was thought through and/or tested.
* Speaking of testing, what's the plan around CI/integration testing of this functionality? Will we enable it by default to check for regressions? How will we prevent regressions with regular kube-proxy integration?

Overall I'm pretty happy with these changes and wouldn't mind if we push more towards getting the base implementation merged then iterate on top to improve it.

borkmann committed:
daemon: allow to define a custom nodeport range

Usage example for defining min/max ports with --node-port-range:

  # ./daemon/cilium-agent --ipv4-range 10.11.0.0/16 --ipv6-range 2001:db8:3::/96 --kvstore-opt consul.address=192.168.1.125:8500 --kvstore consul --device=eno1 --enable-ipv4=true --enable-ipv6=true --disable-envoy-version-check=true --masquerade=false --tunnel=disabled --auto-direct-node-routes=true --enable-node-port=true --node-port-range=30000,34000

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
daemon: mark host reachable services as beta

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: various minor nodeport improvements

Propagate error codes from node port handling and add a comment for
SNAT code's force_range parameter. Reduce variable scope in CT lookup
and avoid extra lookup when node port is not configured.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer commented on daemon/ipam.go:
```diff
@@ -292,6 +292,10 @@ func (d *Daemon) allocateIPs() error {
 		}
 	}
+	if option.Config.EnableNodePort {
+		log.Infof("  NodePort-Range: %d-%d", option.Config.NodePortMin, option.Config.NodePortMax)
```
nit: This option (and all config options) is already printed to the log upon startup, though you would want to embed the default into the CLI argument declaration [1].

borkmann replied:
Makes sense, I'll do that. Thanks for the hint below!

joestringer commented on daemon/daemon_main.go:
```diff
@@ -615,6 +615,12 @@ func init() {
+	flags.Bool(option.EnableNodePort, false, "Enable NodePort type services by Cilium (beta)")
+	option.BindEnv(option.EnableNodePort)
+
+	flags.StringSlice(option.NodePortRange, []string{}, fmt.Sprintf("Set the min/max NodePort port range (default: %d,%d)", option.NodePortMinDefault, option.NodePortMaxDefault))
```
[1]
```suggestion
	flags.StringSlice(option.NodePortRange, []string{fmt.Sprintf("%d", option.NodePortMinDefault), fmt.Sprintf("%d", option.NodePortMaxDefault)}, "Set the min/max NodePort port range")
```

joestringer reviewed (approved):
Minor nits on latest changes, and should be rebased against the (now merged) iptables policy fix.

borkmann committed:
bpf: refine wild card lookup for node port services from host

Commit 4f4ed24f25da ("bpf: wild card lookup for node port services
from host") has the limitation that we'd do a wild card lookup for
any IP address in the given range. This is really suboptimal as it
should only be performed for IPs in our Cilium managed cluster. One
idea to overcome this limitation is to reuse the ipcache and only
do the wildcard lookup for HOST id entries and loopback addresses.
All the other node port service addresses are plugged in via k8s
event handling anyway. This also solves the issue that we don't
need to do extra book-keeping only for cgroups hook case on all
local addresses. It's also a much better solution than blindly
allowing all private address ranges.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
